clippy_lints/src/multiple_unsafe_ops_per_block.rs

```diff
@@ -126,6 +127,26 @@ impl<'tcx> Visitor<'tcx> for UnsafeExprCollector<'tcx> {
                 return self.visit_expr(e);
             },
 
+            // Do not recurse inside an inner `unsafe` block, it will be checked on its own
+            ExprKind::Block(block, _) if matches!(block.rules, BlockCheckMode::UnsafeBlock(_)) => return,
+
+            // Macro calls containing at least one unsafe operations count
+            // as one unsafe operation.
+            _ if expr.span.from_expansion() => {
+                let mut inner_collector = Self {
+                    tcx: self.tcx,
+                    typeck_results: self.typeck_results,
+                    unsafe_ops: vec![],
+                };
+                walk_expr(&mut inner_collector, expr);
+                if !inner_collector.unsafe_ops.is_empty() {
+                    let span = expr.span.source_callsite();
+                    self.unsafe_ops
+                        .push(("macro call expanded to unsafe operation here", span));
+                }
+                return;
+            },
```

Comment:
This doesn't handle macro arguments properly.

```rust
print!("{}, {}", foo(), foo());
```
This has two unsafe operations added by the user, but this would currently be detected as one.

---

What you'll need to do is extend the visitor store what context the unsafe block is in. When detecting an unsafe operation in a different context walk the call chain up to the block's context and use the immediate callee's context as the source of the unsafe call with each callee counting for at most one unsafe operation. 

If you can't walk to the block's context from the unsafe op then it's a macro argument. Not really sure the best way to count that.

when finding an unsafe operation, walk up the call chain to the unsafe block currently being checked.

Reply:
Since we don't intend to lint the code inside macros themselves, it can even be made simpler by looking for the root context only. Don't hesitate to tell me if you see uncovered cases here.

I've used call spans to distinguish between macro call sites. The "weird" case is when a macro expands into one of its unsafe argument + other unsafe calls, see the test marked "confusing situation".